docs/src/checkpointer.md

```diff
@@ -1,12 +1,136 @@
+the coupled simulation and one should be able to restart a coupled simulation
+just by using the states of the component models. Unfortunately, this is
+currently not the case in `ClimaCoupler`. The main reason for this is the
+complex interdependencies between component models and within `ClimaAtmos` which
+make the initialization step inconsistent. For example, in a coupled simulation,
+the surface albedo should be determined by the surface models and used by the
+atmospheric model for radiation transfer, but `ClimaAtmos` also tries to set the
+surface albedo (since it has to do so when run in standalone mode). In addition
+to this, `ClimaAtmos` has a large cache that has internal interdependencies that
+are hard to disentangle, and changing a field might require changing some other
+field in a different part of the cache. As a result, it is not easy for
+`ClimaCoupler` to consistently do initialization from a cold state. To conclude,
+restarting a simulation exclusively using the states of the component models is
+currently impossible.
+
+Given that restarting a simulation from the state is impossible, `ClimaCoupler`
+needs to save the states and the caches. Let us review how we use
+`ClimaCore.InputOutput` and `JLD2` package to accomplish this.
+
+`ClimaCore.InputOutput` provides a loss-less way to save the content of certain
+`ClimaCore` objects to HDF5 files. Objects saved in this way are not tied to a
+particular computing device or configuration. When running with MPI,
+`ClimaCore.InputOutput` are also efficiently written in parallel.
+
+Unfortunately, `ClimaCore.InputOutput` only supports certain objects, such as
+`Field`s and `Space`s, but the cache in component models is more complex than
+this and contains complex objects with highly stateful quantities (e.g., C
+pointers). Because of this, model states are saved to HDF5 but caches must be
+saved to JLD2 files.
+
+`JLD2` allows us to save more complex objects without writing specific
+serialization methods for every struct. `JLD2` allows us to take a big step
+forward, but there are still several challenges that need to be solved:
+1. `JLD2` does not support CUDA natively. To go around this, we have to move
+  everything onto the CPU first. Then, when the data is read back, we have to
+  move it back to the GPU.
+2. `JLD2` does not support MPI natively. To go around this, each process writes
+  its `jld2` checkpoint and reads it back. This introduces the constraint that
+  the number of MPI processes cannot change across restarts.
+3. Some quantities are best not saved and read (for example, anything with
+  pointers). For this, we write a recursive function that traverses the cache
+  and only restores quantities of a certain type (typically, `ClimaCore`
+  objects)
+
+Point 3. adds significant amount of code and requires component models to
+specify how their cache has to be restored.
+
+If you are adding a component model with a cache, you have to extend the
```

Comment:
If a user adds a component model and wants to checkpoint, they'll also need to extend `Checkpointer.get_model_prog_state`. I know this was an existing function before this PR, but could you mention this in the docs please?